File: services/client/src/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"fmt"
	"io"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

type MessageType byte

const (
	Hello   MessageType = 'H'
	Bet     MessageType = 'B'
	Done    MessageType = 'D'
	Winners MessageType = 'W'
)

const messageTypeSize = 1
const payloadLengthSize = 4
const headerSize = messageTypeSize + payloadLengthSize

func WriteMessage(conn io.Writer, messageType MessageType, payload []byte) error {
	header := make([]byte, headerSize)
	header[0] = byte(messageType)
	binary.BigEndian.PutUint32(header[messageTypeSize:], uint32(len(payload)))

	if err := safe_socket.SendAll(conn, append(header, payload...)); err != nil {
		return fmt.Errorf("sending message: %w", err)
	}
	return nil
}

func ReadMessage(conn io.Reader) (MessageType, []byte, error) {
	header, err := safe_socket.RecvAll(conn, headerSize)
	if err != nil {
		return 0, nil, fmt.Errorf("reading message header: %w", err)
	}

	messageType := MessageType(header[0])
	payloadLength := binary.BigEndian.Uint32(header[messageTypeSize:])

	payload, err := safe_socket.RecvAll(conn, int(payloadLength))
	if err != nil {
		return 0, nil, fmt.Errorf("reading message payload: %w", err)
	}

	return messageType, payload, nil
}

File: services/server/src/protocol/protocol.py
import socket
import safe_socket

HELLO = b"H"
BET = b"B"
DONE = b"D"
WINNERS = b"W"
ACK = b"A"

_MESSAGE_TYPE_SIZE = 1
_PAYLOAD_LENGTH_SIZE = 4


def write_message(sock: socket.socket, message_type: bytes, payload: bytes) -> None:
    header = message_type + len(payload).to_bytes(_PAYLOAD_LENGTH_SIZE, "big")
    safe_socket.send_all(sock, header + payload)


def read_message(sock: socket.socket) -> tuple[bytes, bytes]:
    header = safe_socket.recv_all(
        sock, _MESSAGE_TYPE_SIZE + _PAYLOAD_LENGTH_SIZE
    )
    if not header:
        # Peer closed the connection instead of sending a new message
        raise ConnectionError("connection closed before a new message header")

    message_type = header[:_MESSAGE_TYPE_SIZE]
    payload_length = int.from_bytes(header[_MESSAGE_TYPE_SIZE:], "big")
    payload = safe_socket.recv_all(sock, payload_length)
    return message_type, payload
